core/extensions/impl/storage_extension.cpp

```diff
@@ -3,99 +3,188 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#include <exception>
-
 #include "extensions/impl/storage_extension.hpp"
 
 namespace kagome::extensions {
-  uint8_t *StorageExtension::ext_child_storage_root(
-      const uint8_t *storage_key_data, uint32_t storage_key_length,
-      uint32_t *written) {
-    std::terminate();
+  StorageExtension::StorageExtension(
+      std::shared_ptr<storage::merkle::TrieDb> db,
+      std::shared_ptr<runtime::WasmMemory> memory,
+      std::shared_ptr<storage::merkle::Codec> codec, common::Logger logger)
+      : db_(std::move(db)),
+        memory_(std::move(memory)),
+        codec_(std::move(codec)),
+        logger_(std::move(logger)) {}
+
+  // -------------------------Data storage--------------------------
+
+  void StorageExtension::ext_clear_prefix(runtime::WasmPointer prefix_data,
+                                          runtime::SizeType prefix_length) {
+    auto prefix = memory_->loadN(prefix_data, prefix_length);
+    db_->clearPrefix(prefix);
   }
 
-  void StorageExtension::ext_clear_child_storage(
-      const uint8_t *storage_key_data, uint32_t storage_key_length,
-      const uint8_t *key_data, uint32_t key_length) {
-    std::terminate();
+  void StorageExtension::ext_clear_storage(runtime::WasmPointer key_data,
+                                           runtime::SizeType key_length) {
+    auto key = memory_->loadN(key_data, key_length);
+    db_->del(key);
   }
 
-  void StorageExtension::ext_clear_prefix(const uint8_t *prefix_data,
-                                          uint32_t prefix_length) {
-    std::terminate();
+  runtime::SizeType StorageExtension::ext_exists_storage(
+      runtime::WasmPointer key_data, runtime::SizeType key_length) {
+    auto key = memory_->loadN(key_data, key_length);
+    return db_->contains(key) ? 1 : 0;
   }
 
-  void StorageExtension::ext_clear_storage(const uint8_t *key_data,
-                                           uint32_t key_length) {
-    std::terminate();
+  runtime::WasmPointer StorageExtension::ext_get_allocated_storage(
+      runtime::WasmPointer key_data, runtime::SizeType key_length,
+      runtime::WasmPointer len_ptr) {
+    auto key = memory_->loadN(key_data, key_length);
+    auto data = db_->get(key);
+    const auto length = data.has_value() ? data.value().size()
+                                         : runtime::WasmMemory::kMaxMemorySize;
+    memory_->store32(len_ptr, length);
+
+    if (not data) {
+      return 0;
+    }
+
+    auto data_ptr = memory_->allocate(length);
+
+    if (data_ptr != -1) {
+      memory_->storeBuffer(data_ptr, data.value());
+    }
+    return data_ptr;
```

Comment:
Is there sense in returning a pointer, if the allocate fails?

Reply:
I guess client code should take that into account. I will add comment in doc about that and add logging if such issue occurs